src/target/riscv/riscv-013.c

```diff
@@ -3109,47 +3102,69 @@ static int read_sbcs_nonbusy(struct target *target, uint32_t *sbcs)
 }
 
 /* TODO: return mem_access_result_t */
-static int modify_privilege(struct target *target, uint64_t *mstatus, uint64_t *mstatus_old)
+static int modify_privilege(struct target *target, riscv_reg_t *mstatus, riscv_reg_t *mstatus_old,
+		riscv_reg_t *dcsr, riscv_reg_t *dcsr_old)
 {
 	assert(mstatus);
 	assert(mstatus_old);
+	assert(dcsr);
+	assert(dcsr_old);
 	if (!riscv_virt2phys_mode_is_hw(target))
 		return ERROR_OK;
 
-	/* TODO: handle error in this case
-	 * modify_privilege function used only for program buffer memory access.
-	 * Privilege modification requires progbuf size to be at least 5 */
-	if (!has_sufficient_progbuf(target, 5)) {
-		LOG_TARGET_WARNING(target, "Can't modify privilege to provide "
-				"hardware translation: program buffer too small.");
-		return ERROR_OK;
-	}
-
-	/* Read DCSR */
-	riscv_reg_t dcsr;
-	if (register_read_direct(target, &dcsr, GDB_REGNO_DCSR) != ERROR_OK)
+	/* Read and save DCSR */
+	if (riscv_reg_get(target, dcsr, GDB_REGNO_DCSR) != ERROR_OK)
 		return ERROR_FAIL;
+	*dcsr_old = *dcsr;
 
 	/* Read and save MSTATUS */
-	if (register_read_direct(target, mstatus, GDB_REGNO_MSTATUS) != ERROR_OK)
+	if (riscv_reg_get(target, mstatus, GDB_REGNO_MSTATUS) != ERROR_OK)
 		return ERROR_FAIL;
 	*mstatus_old = *mstatus;
 
 	/* If we come from m-mode with mprv set, we want to keep mpp */
-	if (get_field(dcsr, CSR_DCSR_PRV) == PRV_M)
+	if (get_field(*dcsr, CSR_DCSR_PRV) == PRV_M)
 		return ERROR_OK;
 
 	/* mstatus.mpp <- dcsr.prv */
-	*mstatus = set_field(*mstatus, MSTATUS_MPP, get_field(dcsr, CSR_DCSR_PRV));
+	*mstatus = set_field(*mstatus, MSTATUS_MPP, get_field(*dcsr, CSR_DCSR_PRV));
 
 	/* mstatus.mprv <- 1 */
 	*mstatus = set_field(*mstatus, MSTATUS_MPRV, 1);
 
 	/* Write MSTATUS */
-	if (*mstatus == *mstatus_old)
+	if (*mstatus != *mstatus_old &&
+			riscv_reg_set(target, GDB_REGNO_MSTATUS, *mstatus) != ERROR_OK)
+		return ERROR_FAIL;
+
+	/* dcsr.mprven <- 1 */
+	*dcsr = set_field(*dcsr, CSR_DCSR_MPRVEN, CSR_DCSR_MPRVEN_ENABLED);
+
+	/* Write DCSR */
+	if (*dcsr != *dcsr_old &&
+			riscv_reg_set(target, GDB_REGNO_DCSR, *dcsr) != ERROR_OK)
+		return ERROR_FAIL;
+
+	return ERROR_OK;
+}
+
+static int restore_privilege(struct target *target, riscv_reg_t mstatus, riscv_reg_t mstatus_old,
```

Comment:
```suggestion
static int restore_privilege_from_virt2phys_mode(struct target *target, riscv_reg_t mstatus, riscv_reg_t mstatus_old,
```

If you change the name of `modify_privilege()`, then change it here as well.

Reply:
Addressed

Reply:
I feel like `restore_privilege_for_virt2phys_mode()` sounds wrong as we are restoring it from it. I would either remove the `for` word or replace it with `from`.

Thanks

Reply:
Addressed